### ISSUE TITLE
Cannot access packaged files after memory growth

### ISSUE THREAD

**Issue by kichikuou** [kripken/emscripten#5179]
test.c:

    #include <stdio.h>
    #include <stdlib.h>
    
    main() {
      FILE *fp;
      int c;
    
      malloc(20000000);  // Enlarge memory
    
      fp = fopen("test.c", "r");
      while ((c = fgetc(fp)) != EOF)
        putchar(c);
    }

`$ emcc -s WASM=1 -s ALLOW_MEMORY_GROWTH=1 --preload-file test.c -o test.html test.c`

This throws "TypeError: attempting to access detached ArrayBuffer" when attempted to read from test.c.

It seems [packaged files are stored in the heap](https://github.com/kripken/emscripten/blob/b019ab87ef671b5cd6b2221b7fe70dc3eac296ed/tools/file_packager.py#L513). Probably memfs continues to hold a reference to the old heap after the memory grows?

Emscripten 1.37.9 / Firefox 53.0

**Comment by kripken** [comment]
Yes, that's correct, it does avoid copies, but looks like we didn't realize that doesn't mix well with memory growth.

It's possible that building with `--no-heap-copy` would make this work, in which case, we might want to just auto-enable that for memory growth.

**Comment by olafmaas** [comment]
Just ran into the same problem. Building with ```--no-heap-copy``` solved the problem for me. 

Chrome will output  ```Cannot perform TypedArrayCreate on a detached ArrayBuffer``` (Just mentioning this to make searching a bit more easy for other Chrome users with te same problem. :stuck_out_tongue: )

**Comment by nstoddard** [comment]
I just ran into this, and adding --no-heap-copy fixes it for me. Any chance this can be enabled by default when memory growth is enabled?

**Comment by juj** [comment]
Very good find. I'm a bit surprised as to why I haven't run into this before, even though I've used these in conjunction a lot. Though I think I may have been passing a maximum size always, which probably reserves the address space up front and does not end up detaching the buffer, so this issue is avoided then.

Proposed fix in above PR to use --no-heap-copy in that scenario.

**Comment by ztl8702** [comment]
I encountered a situation where even with `--no-heap-copy` enabled this error can still happen.

I was trying to port [`librime`](https://github.com/rime/librime), a C++ library to WebAssembly. In my particular use case, because the code was never optimised for the browser, it used a lot of memory (> 128MB). I initially compiled it with `-s ALLOW_MEMORY_GROWTH=1` and observed the memory size grew 4 times, from 16MB to 256MB, doubling each time.  `TypeError: attempting to access detached ArrayBuffer` occurred after the last growth.

The codebase made use of `Boost.Interprocess`, and around the time that `TypeError` was thrown it was attempting to create a file mapping using Boost. The related code is here:

https://github.com/rime/librime/blob/8d8d2e6c2725041552bb17c787978cf319460d44/src/rime/dict/mapped_file.cc#L54

Below is part of the logs when the error happened, I am looking for ways to reproduce this with in simpler codebase.

![screenshot from 2018-03-18 03-16-50](https://user-images.githubusercontent.com/2714446/37557592-e155c1e8-2a5a-11e8-8f11-75596fe3dc9f.png)

**Interestingly**, everything works fine when I added `-s TOTAL_MEMORY=256MB` to the compile options. So I think it likely has to do with memory growth.

**Comment by meldron** [comment]
Hey guys,

so I had the same problem as @ztl8702 with opus/speedexdsp.

The third memory growth failed:

```
enlarged memory arrays from 16777216 to 33554432, took 26 ms (has ArrayBuffer.transfer? false)
enlarged memory arrays from 33554432 to 67108864, took 67 ms (has ArrayBuffer.transfer? false)
OpusEncoderWorker.js:1 Uncaught TypeError: Cannot perform %TypedArray%.prototype.set on a detached ArrayBuffer

Software Versions.
```
emcc 1.37.33 & 1.37.39
Chrome 66.0.3359.139
Ubuntu 16.04
```

```

So if I used `-s TOTAL_MEMORY=256MB`, everything works fine.

Are there any drawbacks for using such a big `TOTAL_MEMORY`?

**Comment by Inreal** [comment]
Same problem here. I get 

> "TypeError: Cannot perform Construct on a detached ArrayBuffer"

 with `ALLOW_MEMORY_GROWTH`,  `--no-heap-copy` does not change anything. It occurs on a file read operation on a packaged file. Any idea of what I could try?

**Comment by kripken** [comment]
@Inreal The stack trace might help figure that out - the question is where a reference to a detached ArrayBuffer has remained.

**Comment by kripken** [comment]
Thanks, I think the interesting part is at `at Object.read (enscape.js:2633)`. What function is on that line?

(Also maybe `at Object.read (enscape.js:4343)` matters too.)

**Comment by Inreal** [comment]
Ok, these ones are the code snippets:
<img width="507" alt="2633" src="https://user-images.githubusercontent.com/7074079/45104176-c61b6c80-b131-11e8-8473-388df633957d.PNG">

Line 4343 links to that one:
<img width="510" alt="4343" src="https://user-images.githubusercontent.com/7074079/45104202-d29fc500-b131-11e8-913d-50be821e20c9.PNG">

Thanks for having a look @kripken :)

**Comment by kripken** [comment]
Thanks @Inreal. My best guess is that we keep a reference to the main buffer through the `canOwn` parameter being true in a call to `write` in `library_memfs.js`. We should probably not allow that to be true in memory growth.

I'll run the test suite to see if we have coverage for that code path - if not, we'll need to create a testcase for it. If you have a simple one (in the format of the existing tests in the test suite) that would be very useful.

**Comment by kripken** [comment]
There was indeed a relevant test. I extended it and opened a PR with a fix, #7142.
